### DISCUSSION TITLE
Remix Vite: How to configure build target for client files vs server files?

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [remix-run/remix#8856] [Q&A]
## Question
Is it possible to have a build target for the server bundle and another one for the client bundle?

## Examples Use Cases:
Want to use modern JS for server & older style JS output for browser bundles to support older browsers

Keep top-await during the startup (server), and I don’t want to set target: esnext for the browser

There is a `serverModuleFormat` option for server build, but I'm having trouble finding options for client related bundle output:

```ts
export default defineConfig({
  plugins: [
    remix({
      // ⬇️ The output format of the server build. Defaults to "esm".
      serverModuleFormat: "esm",
      ignoredRouteFiles: ["**/.*"],
    }),
    tsconfigPaths(),
  ],
  build: {
    target: "ES2022",
  },
});
```

The `build` option you can pass to the top level object of `defineConfig` is confusing and raises questions like:
- why can we specify `serverModuleFormat` inside of `remix` vite plugin but not something like `browserModuleFormat`?
    - now that builds get split into `build/{client, server}` would be nice to be able to target the build target directly for the final `server` and `client` bundles inside the final build folder
- does `serverModuleFormat` override `build` or is it the other way around?

Create fresh remix app command:
```sh
npx create-remix --template remix-run/remix/templates/vite-express
```

### Related Discussions
- https://github.com/remix-run/remix/issues/7969#issuecomment-1807409678

**Comment by cliffordfajardo** [comment]
Suggestion from remix [discord](https://rmx.as/discord)
> have you tried to use different tsconfigs and bundle the project 2x somehow? not sure if that's possible
> - [remix discord thread](https://discord.com/channels/770287896669978684/1210201098493632583/1210233435193614377)

I would need to try this out sometime, but leaving this comment here in case someone else needed ideas too

**Comment by sergiodxa** [comment]
> why can we specify serverModuleFormat inside of remix vite plugin but not something like browserModuleFormat?

The browser doesn't have CJS support, so the build always uses ESM, what else would you want to configure there?

Meanwhile, Node can be used with CJS or ESM, so depending on your Node version and your HTTP server code you may need the server build to be CJS or ESM, that's why it's an option.

> now that builds get split into build/{client, server} would be nice to be able to target the build target directly for the final server and client bundles inside the final build folder

Build has always been split between client and server, before you had `public/build` for client and `build` for server, the only difference now is they're both in a build folder and inside a client or server folder there.

> does serverModuleFormat override build or is it the other way around?

It only configure the modules to use CJS (`require`) vs ESM (`import`).

Note that `serverModuleFormat` has nothing to do with the target JS version, targeting ES2015 for example means you can't use newer things like async/await or String#replaceAll, and instead the build needs to polypill them and use a fallback (e.g. replace async/await with Promise chains).

You can target ES2023 if you want and still use CJS in Node as the module format.

**Reply by SahidMiller** [reply]
I think the main question is whether Vite will target es2022 for both client and server when setting the build target to es2022. If so, is there any way to set server build target independently of client (for ex. when wanting to use top-level await in the server code)

**Comment by luzat** [comment]
@cliffordfajardo Switching the module format might be harder, but switching the target version for server and client is possible. I am using this to get top-level await only on the server:

```
export default defineConfig(({ isSsrBuild }) => ({
  plugins: [
    remix({
      ignoredRouteFiles: ['**/*'],
      routes: async (defineRoutes) => flatRoutes('routes', defineRoutes),
    }),
    tsconfigPaths(),
  ],
  build: isSsrBuild ? { target: 'ES2022' } : {},
  server: {
    port: 3002,
  },
}))
```
